physlr/physlr.py

```diff
@@ -430,6 +430,54 @@ def determine_backbones(g):
         print(int(timeit.default_timer() - t0), "Determined the backbone paths", file=sys.stderr)
         return backbones
 
+    @staticmethod
+    def wrap_up_messages_and_pass(mst, messages, sender, receiver):
+        """
+        Wrap up all incoming messages to this node (sender) except the one from receiver;
+        and set (pass) the message from sender to receiver.
+        """
+        messages[(receiver, sender)] = \
+            1 + sum([messages[(sender, u)]
+                     for u in mst.neighbors(sender) if u != receiver])
+
+    @staticmethod
+    def determine_reachability_by_message_passing(mst):
+        """
+        Using message passing, determine for each edge of each vertex
+        the number of vertices of the tree reachable from that vertex through that edge.
+        """
+        dfs = list(nx.dfs_edges(mst))
+        if not dfs:
+            return {}
+        stack = [dfs[0][0]]
+        messages = {}
+        # Gather
+        for edge in dfs:
+            while stack[-1] != edge[0]:
+                Physlr.wrap_up_messages_and_pass(mst, messages, stack.pop(), stack[-1])
+            stack.append(edge[1])
+        while len(stack) != 1:
+            Physlr.wrap_up_messages_and_pass(mst, messages, stack.pop(), stack[-1])
+        # Distribute
+        for edge in dfs:
+            Physlr.wrap_up_messages_and_pass(mst, messages, edge[0], edge[1])
+        return messages
+
+    @staticmethod
+    def prune_branches_of_tree(gmst, g, messages, branch_size=50):
+        """"
+        Determine the backbones of the maximum spanning trees
+        and remove branches smaller than branch_size.
+        """
+        g = g.copy()
+        set_of_nodes_for_pruning = {neighbor
+                                    for node in list(g.nodes)
+                                    for neighbor in g.neighbors(node)
```

Comment:
Use `for (u, v) in g.edges()` rather than these two lines.
```suggestion
                                    for u, v in g.edges()
```